File: server/app/utils/validators.py
"""
Validation utility functions.
"""
import re
from datetime import datetime, date


def validate_email(email):
    """
    Validate email format.
    
    Args:
        email: Email string to validate
    
    Returns:
        bool: True if valid, False otherwise
    """
    if not email:
        return False
    
    pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
    return re.match(pattern, email) is not None


def validate_password_strength(password):
    """
    Validate password strength.
    
    Args:
        password: Password string to validate
    
    Returns:
        tuple: (bool, str) - (is_valid, error_message)
    """
    if not password:
        return False, "Password is required"
    
    if len(password) < 6:
        return False, "Password must be at least 6 characters long"
    
    # Optional: Add more complexity requirements
    # if not re.search(r'[A-Z]', password):
    #     return False, "Password must contain at least one uppercase letter"
    
    # if not re.search(r'[0-9]', password):
    #     return False, "Password must contain at least one digit"
    
    return True, ""


def validate_date_of_birth(dob):
    """
    Validate date of birth (must be at least 18 years old).
    
    Args:
        dob: Date object or string in ISO format
    
    Returns:
        tuple: (bool, str) - (is_valid, error_message)
    """
    if isinstance(dob, str):
        try:
            dob = date.fromisoformat(dob)
        except ValueError:
            return False, "Invalid date format. Use YYYY-MM-DD"
    
    if not isinstance(dob, date):
        return False, "Date of birth must be a valid date"
    
    today = date.today()
    age = today.year - dob.year - ((today.month, today.day) < (dob.month, dob.day))
    
    if age < 18:
        return False, "User must be at least 18 years old"
    
    if age > 120:
        return False, "Invalid date of birth"
    
    return True, ""


def validate_gender(gender):
    """
    Validate gender value.
    
    Args:
        gender: Gender string
    
    Returns:
        bool: True if valid, False otherwise
    """
    return gender in ['M', 'F', 'Other']


def validate_role(role):
    """
    Validate user role.
    
    Args:
        role: Role string
    
    Returns:
        bool: True if valid, False otherwise
    """
    return role in ['KORISNIK', 'MENADŽER', 'ADMINISTRATOR']


def validate_amount(amount):
    """
    Validate monetary amount.
    
    Args:
        amount: Amount to validate
    
    Returns:
        tuple: (bool, str) - (is_valid, error_message)
    """
    try:
        amount = float(amount)
        if amount <= 0:
            return False, "Amount must be greater than 0"
        if amount > 1000000:
            return False, "Amount is too large"
        return True, ""
    except (ValueError, TypeError):
        return False, "Invalid amount format"


def sanitize_string(text, max_length=None):
    """
    Sanitize string input (remove extra whitespace, trim).
    
    Args:
        text: Text to sanitize
        max_length: Optional maximum length
    
    Returns:
        str: Sanitized text
    """
    if not text:
        return ""
    
    # Remove extra whitespace
    text = ' '.join(text.split())
    
    # Trim to max length if specified
    if max_length and len(text) > max_length:
        text = text[:max_length]
    
    return text.strip()

File: client/src/utils/validators.js
/**
 * Validate email format
 */
export const validateEmail = (email) => {
  const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
  return emailRegex.test(email);
};

/**
 * Validate password strength
 */
export const validatePassword = (password) => {
  if (!password || password.length < 6) {
    return { valid: false, message: 'Password must be at least 6 characters long' };
  }
  return { valid: true };
};

/**
 * Validate required field
 */
export const validateRequired = (value, fieldName) => {
  if (!value || (typeof value === 'string' && value.trim().length === 0)) {
    return { valid: false, message: `${fieldName} is required` };
  }
  return { valid: true };
};

/**
 * Validate date of birth (must be 18+)
 */
export const validateDateOfBirth = (dob) => {
  if (!dob) {
    return { valid: false, message: 'Date of birth is required' };
  }

  const today = new Date();
  const birthDate = new Date(dob);
  let age = today.getFullYear() - birthDate.getFullYear();
  const monthDiff = today.getMonth() - birthDate.getMonth();
  
  if (monthDiff < 0 || (monthDiff === 0 && today.getDate() < birthDate.getDate())) {
    age--;
  }

  if (age < 18) {
    return { valid: false, message: 'You must be at least 18 years old' };
  }

  return { valid: true };
};

/**
 * Validate positive number
 */
export const validatePositiveNumber = (value, fieldName) => {
  const num = Number(value);
  if (isNaN(num) || num <= 0) {
    return { valid: false, message: `${fieldName} must be a positive number` };
  }
  return { valid: true };
};

/**
 * Validate future date
 */
export const validateFutureDate = (dateString, fieldName) => {
  const date = new Date(dateString);
  const now = new Date();
  
  if (date <= now) {
    return { valid: false, message: `${fieldName} must be in the future` };
  }
  
  return { valid: true };
};

/**
 * Validate rating (1-5)
 */
export const validateRating = (rating) => {
  const num = Number(rating);
  if (isNaN(num) || num < 1 || num > 5) {
    return { valid: false, message: 'Rating must be between 1 and 5' };
  }
  return { valid: true };
};